src/aspire/source/image.py

```diff
@@ -1371,6 +1429,138 @@ def __repr__(self):
+    """
+
+    def __init__(self, src, orientation_estimator=None):
+        """
+        Constructor of an oriented ImageSource object. Orientation is determined by
+        performing orientation estimation using a supplied `orientation_estimator`.
+
+        :param src: Source used for orientation estimation
+        :param orientation_estimator: CLOrient3D subclass used for orientation estimation.
+            Default uses the CLSyncVoting method.
+        """
+
+        self.src = src
+        if not isinstance(self.src, ImageSource):
+            raise ValueError(
+                f"`src` should be subclass of `ImageSource`, found {self.src}."
+            )
+
+        # `indices` for IndexedSource.
+        indices = np.arange(self.src.n)
+
+        super().__init__(
+            src=self.src,
+            indices=indices,
+        )
+
+        # Remove any orientation information passed in by original source.
+        self._reset_orientation()
+
+        # Flag for lazy eval of orientation estimation.
+        self._oriented = False
+
+        if orientation_estimator is None:
+            orientation_estimator = CLSyncVoting(src)
+
+        self.orientation_estimator = orientation_estimator
+        if not isinstance(self.orientation_estimator, CLOrient3D):
+            raise ValueError(
+                "`orientation_estimator` should be subclass of `CLOrient3D`,"
+                f" found {self.orientation_estimator}."
+            )
+
+        # Any further operations should not mutate this instance.
+        self._mutable = False
+
+    def _orient(self):
+        """
+        Perform orientation estimation if not already done.
+        """
+
+        # Short circuit.
+        if self._oriented:
+            logger.debug(f"{self.__class__.__name__} already oriented, skipping")
+            return
+
+        logger.info(
+            f"Estimating rotations for {self.src} using {self.orientation_estimator}."
+        )
+        self.orientation_estimator.estimate_rotations()
+
+        # Allow mutability to set rotations.
+        self._mutable = True
+        self.rotations = self.orientation_estimator.rotations
+        self._mutable = False
+
+        self._oriented = True
+
+    def _reset_orientation(self):
+        """
+        Remove orientation information passed in by original source.
+        """
+        _info_removed = False
+        rot_keys = ["_rlnAngleRot", "_rlnAngleTilt", "_rlnAnglePsi"]
+        for key in rot_keys:
+            if self.has_metadata(key):
+                del self._metadata[key]
+                _info_removed = True
+        if _info_removed:
+            logger.info(f"Removing orientation information passed by {self.src}.")
+
+    def _images(self, indices):
+        """
+        Returns images from `self.src` corresponding to `indices`.
+
+        :param indices: A 1-D NumPy array of indices.
+        :return: An `Image` object.
+        """
+        self._orient()
```

Comment:
Not necessary.

Reply:
hmm, I thought I had this here for a reason. But you are correct. I'll remove the entire `_images()` method from this subclass since we just call super. The only addition was this lazy evaluation.

Reply:
Removed.